xds/internal/balancer/weightedtarget/weightedtarget.go

```diff
@@ -0,0 +1,150 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package weightedtarget implements the weighted_target balancer.
+package weightedtarget
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/hierarchy"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+	"google.golang.org/grpc/xds/internal"
+	"google.golang.org/grpc/xds/internal/balancer/balancergroup"
+)
+
+const weightedTargetName = "weighted_target_experimental"
+
+func init() {
+	balancer.Register(&weightedTargetBB{})
+}
+
+type weightedTargetBB struct{}
+
+func (wt *weightedTargetBB) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	b := &weightedTargetBalancer{}
+	b.logger = grpclog.NewPrefixLogger(loggingPrefix(b))
+	b.bg = balancergroup.New(cc, nil, b.logger)
+	b.bg.Start()
+	b.logger.Infof("Created")
+	return b
+}
+
+func (wt *weightedTargetBB) Name() string {
+	return weightedTargetName
+}
+
+func (wt *weightedTargetBB) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	return parseConfig(c)
+}
+
+type weightedTargetBalancer struct {
+	logger *grpclog.PrefixLogger
+
+	// The keys in BalancerGroup is currently a locality struct, because
+	// balancer group also does load reporting (which is disabled here since
+	// LoadStore isn't set). The keys are all set to Locality{Region: name}
+	// until load reporting is refactor and moved out from balancer group.
+	bg *balancergroup.BalancerGroup
+
+	targets map[string]target
```

Comment:
Is there a reason to not store the config as type `lbConfig`, and compare with existing config and do nothing when the config has not changed?

Reply:
Comparing two `lbConfig` would still require reading key-values in targets.

- If old/new configs are same
  - Compare: read all key/values in both old and new
- If old/new are different
  - Compare: read key/values in both old and new until a diff is found
  - Apply new: read all key/values in both old and new

So compare-and-do-nothing-if-same would just be an extra step but doesn't save anything?